### DISCUSSION TITLE
Version 3 Ideas [tentative]

### DISCUSSION THREAD

**Discussion by lukas-reineke** [lukas-reineke/indent-blankline.nvim#462] [Ideas]
- Expose API for plugins
   - `get_indent_virt_text({buffer}, {start_row}, {start_col}, {end_row}, {end_col})`
   - `set_indent_virt_text`?
- Unit tests
- Use `tabstop` and `shiftwidth` per indent level instead of global for the whole buffer
- Support `vartabstop` https://github.com/lukas-reineke/indent-blankline.nvim/issues/198
- Support irregular indentation https://github.com/lukas-reineke/indent-blankline.nvim/issues/170
- Better defaults
- Remove old options
- Remove vimscript options
- Custom character for end of indent? https://github.com/lukas-reineke/indent-blankline.nvim/issues/419
- Custom lua hooks
   - `is_enabled`
- Remove logic to guess tabs vs spaces, instead calculate everything per character. Lua is fast enough to handle this.

**Comment by Jlll1** [comment]
I love the corner icon idea. I've never seen something like this in any editor, but this could increase the readibility dramatically.

**Comment by yioneko** [comment]
Hi, I found that `:h nvim_set_decoration_provider()` is a great choice to implement indent guides, and is possible to resolve hard issues like https://github.com/lukas-reineke/indent-blankline.nvim/issues/457 https://github.com/lukas-reineke/indent-blankline.nvim/issues/489 and no need to wait for core support.

I think that you may find it an interesting approach, and here is my [POC with minimal features](https://github.com/yioneko/indentlines-decorator/blob/main/indentlines.lua) to show that it is possible and might be somehow useful. Willing to see that this could be added into next version plan.

**Reply by lukas-reineke** [reply]
I agree, I've also started to experiment with this.
Thank you for the PoC, I'll take a look.
